object is not an instance of declaring class com.hazelcast.query.impl.QueryException

if I deploy a war which load a set of objects in cache and then I deploy the war again I get the following error:

```
Jan 3, 2014 5:35:58 PM com.hazelcast.map.operation.PutAllOperation
java.lang.IllegalArgumentException: java.lang.ClassCastException@7617d861
com.hazelcast.query.impl.QueryException: java.lang.IllegalArgumentException: java.lang.ClassCastException@7617d861
    at com.hazelcast.query.impl.QueryEntry.extractViaReflection(QueryEntry.java:96)
    at com.hazelcast.query.impl.QueryEntry.getAttribute(QueryEntry.java:79)
    at com.hazelcast.query.impl.IndexImpl.saveEntryIndex(IndexImpl.java:62)
    at com.hazelcast.query.impl.IndexService.saveEntryIndex(IndexService.java:68)
    at com.hazelcast.map.DefaultRecordStore.saveIndex(DefaultRecordStore.java:880)
    at com.hazelcast.map.DefaultRecordStore.putFromLoad(DefaultRecordStore.java:811)
    at com.hazelcast.map.operation.PutAllOperation.run(PutAllOperation.java:71)
    at com.hazelcast.spi.impl.BasicOperationService.doRunOperation(BasicOperationService.java:359)
    at com.hazelcast.spi.impl.BasicOperationService.access$1000(BasicOperationService.java:74)
    at com.hazelcast.spi.impl.BasicOperationService$LocalOperationProcessor.run(BasicOperationService.java:834)
    at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
    at java.lang.Thread.run(Thread.java:695)
    at com.hazelcast.spi.impl.BasicOperationService$OperationThread.run(BasicOperationService.java:923)
Caused by: java.lang.IllegalArgumentException: java.lang.ClassCastException@7617d861
    at sun.reflect.GeneratedMethodAccessor25.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
    at java.lang.reflect.Method.invoke(Method.java:597)
    at com.hazelcast.query.impl.ReflectionHelper$MethodGetter.getValue(ReflectionHelper.java:180)
    at com.hazelcast.query.impl.ReflectionHelper.extractValue(ReflectionHelper.java:155)
    at com.hazelcast.query.impl.QueryEntry.extractViaReflection(QueryEntry.java:93)
    ... 13 more
```

--- a/hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java
+++ b/hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java
@@ -32,6 +32,7 @@ import static com.hazelcast.query.QueryConstants.THIS_ATTRIBUTE_NAME;
 
 public class ReflectionHelper {
 
+    private final static ClassLoader THIS_CL = ReflectionHelper.class.getClassLoader();
     private final static ConcurrentMap<String, Getter> getterCache = new ConcurrentHashMap<String, Getter>(1000);
 
     public static AttributeType getAttributeType(Class klass) {
@@ -142,7 +143,7 @@ public class ReflectionHelper {
                 parent = localGetter;
             }
             getter = parent;
-            if (!(getter instanceof ThisGetter)) {
+            if (getter.isCacheable()) {
                 getterCache.putIfAbsent(cacheKey, getter);
             }
             return getter;
@@ -165,6 +166,8 @@ public class ReflectionHelper {
         abstract Object getValue(Object obj) throws Exception;
 
         abstract Class getReturnType();
+
+        abstract boolean isCacheable();
     }
 
     static class MethodGetter extends Getter {
@@ -182,6 +185,11 @@ public class ReflectionHelper {
 
         Class getReturnType() {
             return this.method.getReturnType();
+        }
+
+        @Override
+        boolean isCacheable() {
+            return THIS_CL.equals(method.getDeclaringClass().getClassLoader());
         }
 
         @Override
@@ -208,6 +216,11 @@ public class ReflectionHelper {
         }
 
         @Override
+        boolean isCacheable() {
+            return THIS_CL.equals(field.getDeclaringClass().getClassLoader());
+        }
+
+        @Override
         public String toString() {
             return "FieldGetter [parent=" + parent + ", field=" + field + "]";
         }
@@ -229,6 +242,11 @@ public class ReflectionHelper {
         @Override
         Class getReturnType() {
             return this.object.getClass();
+        }
+
+        @Override
+        boolean isCacheable() {
+            return false;
         }
     }
 }

